util/patricia_trie/src/lib.rs

```diff
@@ -49,155 +49,159 @@ mod lookup;
+	// but if we add a `Box<Error>` here we run into issues in the `ethcore` crate:
+	//  "the trait bound `std::error::Error + 'static: std::marker::Send` is not satisfied"
+	// Investigate if using `Box<Error + Send>` would help here (it does compile).
+	// Another potential problem is that the PartialEq, Eq and Clone derives do not work.
+	DecoderError(T, String),
 }
 
-impl fmt::Display for TrieError {
+impl<T> fmt::Display for TrieError<T> where T: std::fmt::Debug {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		match *self {
-			TrieError::InvalidStateRoot(ref root) => write!(f, "Invalid state root: {}", root),
-			TrieError::IncompleteDatabase(ref missing) =>
-				write!(f, "Database missing expected key: {}", missing),
-			TrieError::DecoderError(ref err) =>  write!(f, "Decoding failed with {}", err),
+			TrieError::InvalidStateRoot(ref root) => write!(f, "Invalid state root: {:?}", root),
+			TrieError::IncompleteDatabase(ref missing) => write!(f, "Database missing expected key: {:?}", missing),
+			TrieError::DecoderError(ref hash, ref decoder_err) =>  write!(f, "Decoding failed for hash {:?}; err: {:?}", hash, decoder_err),
 		}
 	}
 }
 
-impl error::Error for TrieError {
+impl<T> error::Error for TrieError<T> where T: std::fmt::Debug {
 	fn description(&self) -> &str {
 		match *self {
 			TrieError::InvalidStateRoot(_) => "Invalid state root",
 			TrieError::IncompleteDatabase(_) => "Incomplete database",
-			TrieError::DecoderError(ref e) => e.description(),
+			TrieError::DecoderError(_, ref err_string) => err_string,
 		}
 	}
 }
 
-impl From<rlp::DecoderError> for Box<TrieError> {
-	fn from(e: rlp::DecoderError) -> Self { Box::new(TrieError::DecoderError(e)) }
-}
-
-/// Trie result type. Boxed to avoid copying around extra space for `H256`s on successful queries.
-pub type Result<T> = ::std::result::Result<T, Box<TrieError>>;
+/// Trie result type. Boxed to avoid copying around extra space for the `Hasher`s `Out`s on successful queries.
+pub type Result<T, U> = ::std::result::Result<T, Box<TrieError<U>>>;
 
-/// Trie-Item type.
-pub type TrieItem<'a> = Result<(Vec<u8>, DBValue)>;
+/// Trie-Item type used for iterators over trie data.
+pub type TrieItem<'a, U> = Result<(Vec<u8>, DBValue), U>;
 
 /// Description of what kind of query will be made to the trie.
 ///
 /// This is implemented for any &mut recorder (where the query will return
 /// a DBValue), any function taking raw bytes (where no recording will be made),
 /// or any tuple of (&mut Recorder, FnOnce(&[u8]))
-pub trait Query {
+pub trait Query<H: Hasher> {
 	/// Output item.
 	type Item;
 
 	/// Decode a byte-slice into the desired item.
-	fn decode(self, &[u8]) -> Self::Item;
+	fn decode(self, data: &[u8]) -> Self::Item;
 
 	/// Record that a node has been passed through.
-	fn record(&mut self, &H256, &[u8], u32) { }
+	#[allow(unused_variables)]
+	fn record(&mut self, hash: &H::Out, data: &[u8], depth: u32) {}
```

Comment:
better just prefix the names with `_` (e.g. `_hash`) rather than silencing the lint